### PR TITLE
Set limits in MerkleBlockPayload (3x)

### PR DESCRIPTION
**Pull request by shargon**
Close https://github.com/neo-project/neo/issues/1375

Ensure that MerkleBlockPayload fit the expected values

### CODE REVIEW

**Comment by shargon** [comment]
> Not sure this is good because we can.change this limit.

Could you explain more your view point?

**Comment by vncoelho** [comment]
In fact, it is just on the P2P layer, right?
Thus, if we modify the parameter it will not affect syncing of the chain as I initially thought.

File: src/neo/Network/P2P/Payloads/MerkleBlockPayload.cs [hunk 1]
```diff
@@ -37,9 +37,9 @@ public static MerkleBlockPayload Create(Block block, BitArray flags)
         public override void Deserialize(BinaryReader reader)
         {
             base.Deserialize(reader);
-            ContentCount = (int)reader.ReadVarInt(int.MaxValue);
-            Hashes = reader.ReadSerializableArray<UInt256>();
-            Flags = reader.ReadVarBytes();
+            ContentCount = (int)reader.ReadVarInt(Block.MaxTransactionsPerBlock);
+            Hashes = reader.ReadSerializableArray<UInt256>(Block.MaxTransactionsPerBlock);
+            Flags = reader.ReadVarBytes(Block.MaxTransactionsPerBlock);
```
Comment (erikzhang) [inline src/neo/Network/P2P/Payloads/MerkleBlockPayload.cs]: I'm not sure if this is enough.

Comment (Tommo-L) [reply src/neo/Network/P2P/Payloads/MerkleBlockPayload.cs]: How about changing it to the following:

```c#
ContentCount = (int)reader.ReadVarInt(Block.MaxTransactionsPerBlock + 1);
Hashes = reader.ReadSerializableArray<UInt256>(Block.MaxTransactionsPerBlock + 1);
Flags = reader.ReadVarBytes(36000);   // 36000 is the limit in FilterLoadPayload.Deserialize
```

Comment (shargon) [reply src/neo/Network/P2P/Payloads/MerkleBlockPayload.cs]: Flags is related to the number of TX
https://github.com/neo-project/neo/blob/0f75741e2ec471235cba99b6ffc6b38b634dd47e/src/neo/Network/P2P/ProtocolHandler.cs#L236

Comment (erikzhang) [reply src/neo/Network/P2P/Payloads/MerkleBlockPayload.cs]: It's a bit array, so I think we should divide it by 8.

File: src/neo/Network/P2P/Payloads/MerkleBlockPayload.cs [hunk 2]
```diff
@@ -37,9 +37,9 @@ public static MerkleBlockPayload Create(Block block, BitArray flags)
         public override void Deserialize(BinaryReader reader)
         {
             base.Deserialize(reader);
-            ContentCount = (int)reader.ReadVarInt(int.MaxValue);
-            Hashes = reader.ReadSerializableArray<UInt256>();
-            Flags = reader.ReadVarBytes();
+            ContentCount = (int)reader.ReadVarInt(Block.MaxTransactionsPerBlock);
+            Hashes = reader.ReadSerializableArray<UInt256>(Block.MaxTransactionsPerBlock);
+            Flags = reader.ReadVarBytes(Block.MaxTransactionsPerBlock / 8);
```
Comment (erikzhang) [inline src/neo/Network/P2P/Payloads/MerkleBlockPayload.cs]:
```suggestion
            Flags = reader.ReadVarBytes((Block.MaxTransactionsPerBlock + 7) / 8);
```

File: src/neo/Network/P2P/Payloads/MerkleBlockPayload.cs (hunk 1 above)
Comment (Tommo-L) [reply src/neo/Network/P2P/Payloads/MerkleBlockPayload.cs]: Why not set `ContentCount = (int)reader.ReadVarInt(Block.MaxTransactionsPerBlock + 1)` and `Hashes = reader.ReadSerializableArray<UInt256>(Block.MaxTransactionsPerBlock + 1)`, as `ContentCount = block.Transactions.Length + 1`

Comment (shargon) [reply src/neo/Network/P2P/Payloads/MerkleBlockPayload.cs]: Seems good to me too

**Pull request merged**